### ISSUE TITLE
Twig Extension doesn't notify when the Url Generator is missing

### ISSUE THREAD

**Issue by niklongstone** [puli/issues#98]
The issue occurs when you want to use the resource_url function in the Puli Twig Extension but you don't have the Url Generator installed.

If you want to use the Puli Twig extension without injecting the Url Generator no exception is thrown.

Line affected is [PuliExtension.php:101](https://github.com/puli/twig-extension/blob/master/src/PuliExtension.php):
```
    public function getFunctions()
    {
        if (!$this->urlGenerator) {
            return array();
        }
```

Code snippet to reproduce the issue:
```
<?php
require '../vendor/autoload.php';

use Puli\TwigExtension\PuliTemplateLoader;
use Puli\TwigExtension\PuliExtension;

$factoryClass = PULI_FACTORY_CLASS;
$factory = new $factoryClass();
$myRepo = $factory->createRepository();

$twig = new \Twig_Environment(new PuliTemplateLoader($myRepo));
$twig->addExtension(new PuliExtension($myRepo));


echo $twig->render('/myapp/views/index.html.twig');
```

**Comment by mnapoli** [comment]
Hi, what is the problem? I'm asking because I'm currently using it without the URL generator too but I don't see any problem (though I admit I have only a basic usage for now). So if no exception is thrown, what is the issue?

**Comment by webmozart** [comment]
@mnapoli I think the problem is that the `resource_url()` function is not available, but it's hard for the user to find out why. I think it's safe to make the URL generator a mandatory dependency, since most people will need that function. What do you think?

**Comment by niklongstone** [comment]
@mnapoli the issue occurs when you want to use the `resource_url` function in the Puli Twig Extension but you don't have the Url Generator installed.

I will edit the issue to make it clear.

Because of this direct dependency. I think it worth at least having an exception to notify the user or  having a "mandatory dependency".

**Comment by mnapoli** [comment]
Thanks for explaining. Making it mandatory is an option but how about throwing an exception? For example:

> The resource_url() Twig function requires a UrlGenerator to be set in PuliExtension

As a user, it feels clear enough.

In my use case I currently don't have a URL generator (because I'm just getting started). I'm not sure whether the URL generator is a central part of Puli (haven't used it yet, but will probably soon)? If it is, then indeed it's better to require it.

**Comment by gossi** [comment]
I'd like to reopen this one, because the error message is irritating. I had the `puli/url-generator` package installed and get the warning about a non-installed url-generator - huh? My problem was I didn't pass the url-generator as second parameter to the `PuliExtension`.

Before: `new PuliExtension($repo)`
After: `new PuliExtension($repo, $discovery)`

Also docs don't mention the second parameter as required for use in conjunction with `resource_url()` twig function. I think you find a smart solution to that :smiley:

**Comment by webmozart** [comment]
@gossi Thanks for the report! Could you open a new issue please?
